lib/babe/build.go

```diff
@@ -175,38 +176,44 @@ func (b *Service) buildBlockBABEPrimaryPreDigest(slot Slot) (*types.BabePrimaryP
 // for each extrinsic in queue, add it to the block, until the slot ends or the block is full.
 // if any extrinsic fails, it returns an empty array and an error.
 func (b *Service) buildBlockExtrinsics(slot Slot) []*transaction.ValidTransaction {
-	next := b.nextReadyExtrinsic()
-	included := []*transaction.ValidTransaction{}
+	var included []*transaction.ValidTransaction
 
-	for !hasSlotEnded(slot) && next != nil {
-		logger.Trace("build block", "applying extrinsic", next)
+	for !hasSlotEnded(slot) {
+		txn := b.transactionState.Pop()
+		// Transaction queue is empty.
+		if txn == nil {
+			return included
+		}
 
-		t := b.transactionState.Pop()
-		ret, err := b.rt.ApplyExtrinsic(next)
-		if err != nil {
-			logger.Warn("failed to apply extrinsic", "error", err, "extrinsic", next)
-			next = b.nextReadyExtrinsic()
+		// Move to next extrinsic.
+		if txn.Extrinsic == nil {
 			continue
 		}
 
-		// if ret == 0x0001, there is a dispatch error; if ret == 0x01, there is an apply error
-		if ret[0] == 1 || bytes.Equal(ret[:2], []byte{0, 1}) {
-			errTxt, err := determineError(ret)
-			// remove invalid extrinsic from queue
-			if err == nil {
-				logger.Warn("failed to interpret extrinsic error", "error", ret, "extrinsic", next)
-			} else {
-				logger.Warn("failed to apply extrinsic", "error", errTxt, "extrinsic", next)
-			}
-
-			next = b.nextReadyExtrinsic()
+		extrinsic := txn.Extrinsic
+		logger.Trace("build block", "applying extrinsic", extrinsic)
+
+		ret, err := b.rt.ApplyExtrinsic(extrinsic)
+		if err != nil {
+			logger.Warn("failed to apply extrinsic", "error", err, "extrinsic", extrinsic)
 			continue
 		}
 
-		logger.Debug("build block applied extrinsic", "extrinsic", next)
+		ok, err := determineErr(ret)
```

Comment:
maybe introduce new error types  and type check `err` for `DispatchOutcome` or `TransactionValidityError` instead of relying on this `bool` return.

Reply:
Done